### PR TITLE
Add support for SonarQube

### PR DESCRIPTION
**Pull request by bbbneo333**
## Description
- Added support for SonarQube

## Database schema changes
- NA

### Automated test cases added
- NA

### Manual test cases run
- NA

### CODE REVIEW

**Review by jjalan — COMMENTED** [review]
To ensure this is working as expected, is it possible to do a commit that violates Sonarqube rule and see this fail?

**Review by jjalan — COMMENTED** [review]
Can we rename the code analysis step in the CI process a bit more readable?

Maybe `ci/analyze`?

Screenshot 2022-02-27 at 9 48 53 AM

**Review by jjalan — COMMENTED** [review]
Semantically, it also feels a bit odd that our CI process would be handled by two system:

1. CircleCI for build, lint, test
2. Github for analyze

It is possible that it is out of scope for this PR but it would be good to use one CI system for all the 4 steps - build, lint, analyze, test

File: .circleci/config.yml [hunk 1]
```diff
@@ -45,6 +45,21 @@ jobs:
...
           command: npm run test
+  analyze:
+    docker:
+      - image: sonarsource/sonar-scanner-cli:latest
+    steps:
+      - checkout
+      - run:
+          name: sonarqube_scan
+          command: >-
+                  sonar-scanner-ee
```
Comment (bbbneo333) [inline .circleci/config.yml:56]: Try `sonar-scanner` ?

Comment (bbbneo333) [reply .circleci/config.yml:56]: Missing `sonar.host.url`

**Comment by jjalan** [comment]
There are two sonarqube entry, not sure if its desirable.

Screenshot 2022-03-01 at 10 31 03 PM

File: .circleci/config.yml [hunk 2]
```diff
@@ -45,6 +46,38 @@ jobs:
...
+            echo 'export CIRCLE_PR_NUMBER=$CIRCLE_PR_NUMBER' >> $BASH_ENV
+      - run:
+          name: SonarQube - PullRequest Scan
+          command: >-
+            sonar-scanner
+            -X
+            -Dsonar.login=$SONAR_TOKEN
+            -Dsonar.host.url=https://sonarqube.platform.jalantechnologies.com
+            -Dsonar.pullrequest.key=20
+            -Dsonar.pullrequest.branch=feature/integrate_sonarqube
```
Comment (bbbneo333) [inline .circleci/config.yml:75]: Replace value here with `$CIRCLE_BRANCH`

File: .circleci/config.yml [hunk 3]
```diff
@@ -45,6 +46,38 @@ jobs:
...
+
+            echo 'export CIRCLE_PR_NUMBER=$CIRCLE_PR_NUMBER' >> $BASH_ENV
+      - run:
+          name: SonarQube - PullRequest Scan
+          command: >-
+            sonar-scanner
+            -X
+            -Dsonar.login=$SONAR_TOKEN
+            -Dsonar.host.url=https://sonarqube.platform.jalantechnologies.com
+            -Dsonar.pullrequest.key=20
```
Comment (bbbneo333) [inline .circleci/config.yml:74]: Replace value here with `$CIRCLE_PR_NUMBER`

File: .circleci/config.yml [hunk 4]
```diff
@@ -45,6 +46,38 @@ jobs:
...
+            fi
+
+            echo 'export CIRCLE_PR_NUMBER=$CIRCLE_PR_NUMBER' >> $BASH_ENV
+      - run:
+          name: SonarQube - PullRequest Scan
+          command: >-
+            sonar-scanner
+            -X
+            -Dsonar.login=$SONAR_TOKEN
+            -Dsonar.host.url=https://sonarqube.platform.jalantechnologies.com
```
Comment (bbbneo333) [inline .circleci/config.yml:73]: Replace value here with `$SONAR_HOST_URL` (Please do check what's the env variable we are using here)

File: .github/workflows/analysis-sonarqube-pull-request.yml [hunk 5]
```diff
@@ -0,0 +1,28 @@
+name: analysis:sonarqube:pull-request
```
Comment (bbbneo333) [inline .github/workflows/analysis-sonarqube-pull-request.yml:1]: Please remove this as we are now running analysis on CircleCi

File: .github/workflows/analysis-sonarqube-push.yml [hunk 6]
```diff
@@ -0,0 +1,26 @@
+name: analysis:sonarqube:push
```
Comment (bbbneo333) [inline .github/workflows/analysis-sonarqube-push.yml:1]: Please remove this as we are now running analysis on CircleCi

**Comment by bbbneo333** [comment]
@hsahu615 - Can you look into why we have an additional check now? ("
boilerplate-node-ts Sonarqube Results"). Shouldn't the results should be directly reported via "ci/circleci: analyze"? I think this has something to do with "sonar.qualitygate.wait=true" property.

Screenshot 2022-03-08 at 12 52 47 PM

**Comment by bbbneo333** [comment]
> To ensure this is working as expected, is it possible to do a commit that violates Sonarqube rule and see this fail?

@hsahu615  - Can you address this one as well please? Thanks!

**Comment by bbbneo333** [comment]
@hsahu615 - We also need to add support for running branch analysis on `main` branch.

File: .circleci/config.yml [hunk 7]
```diff
@@ -45,6 +46,38 @@ jobs:
...
+      - run:
+          name: SonarQube - PullRequest Scan
+          command: >-
+            sonar-scanner
+            -X
+            -Dsonar.login=$SONAR_TOKEN
+            -Dsonar.host.url=$SONAR_HOST_URL
+            -Dsonar.pullrequest.key=${CIRCLE_PULL_REQUEST##*/}
+            -Dsonar.pullrequest.branch=$CIRCLE_BRANCH
+            -Dsonar.pullrequest.base=develop
```
Comment (bbbneo333) [inline .circleci/config.yml:76]: Let's change this to `main`

**Comment by bbbneo333** [comment]
On every push to the main branch, circleci should run branch analysis using the `-Dsonar.branch.name=main` as arguments. Please see this workflow file on how we were doing it using GitHub actions to get the idea - https://github.com/jalantechnologies/jtc-website-v2/blob/develop/.github/workflows/analysis-sonarqube-push.yml

File: .circleci/config.yml [hunk 8]
```diff
@@ -45,6 +46,46 @@ jobs:
...
```
Comment (bbbneo333) [inline .circleci/config.yml:76]: Why was this reverted? Please change it back to `main`

**Pull request merged**